internal/gatewayapi/tls.go

```
@@ -96,26 +98,66 @@ func parseCertsFromTLSSecretsData(secrets []*corev1.Secret) ([]*x509.Certificate
+// Returns an error if no valid certificates are found in the provided data.
+func filterValidCertificates(data []byte) ([]byte, error) {
+	if len(data) == 0 {
+		return nil, fmt.Errorf("no certificate data provided")
 	}
+
 	now := time.Now()
-	for _, cert := range certs {
-		if now.After(cert.NotAfter) {
-			return fmt.Errorf("certificate %s has expired since %v", cert.Subject.CommonName, cert.NotAfter)
+	var validPemCount int
+	var errs []error
+	validData := make([]byte, 0, len(data))
+
+	// Process each PEM block in the data
+	rest := data
+	for len(rest) > 0 {
+		block, remaining := pem.Decode(rest)
+		if block == nil {
+			break
 		}
-		if now.Before(cert.NotBefore) {
-			return fmt.Errorf("certificate %s will be valid after %v", cert.Subject.CommonName, cert.NotBefore)
+		rest = remaining
+
+		// Parse all certificates in this PEM block
+		certs, err := x509.ParseCertificates(block.Bytes)
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+
+		// Validate all certificates in this PEM block
+		blockValid := true
+		for _, cert := range certs {
+			if now.After(cert.NotAfter) {
+				errs = append(errs, fmt.Errorf("certificate %s has expired since %v", cert.Subject.CommonName, cert.NotAfter))
+				blockValid = false
+				break
+			}
+			if now.Before(cert.NotBefore) {
+				errs = append(errs, fmt.Errorf("certificate %s will be valid after %v", cert.Subject.CommonName, cert.NotBefore))
+				blockValid = false
+				break
+			}
+		}
+		// Only include this PEM block if all certificates in it are valid
+		if blockValid {
+			validData = append(validData, pem.EncodeToMemory(block)...)
+			validPemCount++
 		}
 	}
-	return nil
+
+	// Return error if no valid certificates were found
+	if validPemCount == 0 {
+		if len(errs) > 0 {
+			return nil, errors.Join(errs...)
+		}
+		// No errors but no valid PEM blocks found - PEM decoding failed
+		return nil, fmt.Errorf("unable to decode pem data for certificate")
+	}
+	return validData, nil
```

We are not returning the expired cert errors here if there is a valid cert. IMO we should return the error and log it as warning if `validData` is non empty.